src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,233 @@
+  OwnableWithGuardian,
+  Rescuable
+{
+  using SafeERC20 for IERC20;
+
+  /// @notice The Balancer Vault contract
+  IBalancerVault public immutable VAULT;
+
+  /// @notice The number of tokens in the pool
+
+  /// @notice The address of the Hypernative service
+  address public immutable HYPERNATIVE;
+
+  /// @dev Mapping of token configurations (token to provider)
+  mapping(address token => address provider) public override tokenProvider;
+
+  /// @dev Restricts access to only the owner, guardian, or Hypernative address.
+  modifier onlyWithdrawable() {
+    if (_msgSender() != owner() && _msgSender() != guardian() && _msgSender() != HYPERNATIVE) {
+      revert Unauthorized();
+    }
+    _;
+  }
+
+  /**
+   * @notice Constructor to initialize the contract.
+   * @param _vault The address of the Balancer Vault.
+   * @param _tokenConfig The array of token configurations (address and provider).
+   * @param _owner The owner of the contract.
+   * @param _guardian The guardian of the contract.
+   * @param _hypernative The Hypernative service address.
+   */
+  constructor(
+    address _vault,
+    TokenConfig[] memory _tokenConfig,
+    address _owner,
+    address _guardian,
+    address _hypernative
+  ) {
+    VAULT = IBalancerVault(_vault);
+
+    for (uint256 i = 0; i < _tokenConfig.length; ) {
+      tokenProvider[_tokenConfig[i].token] = _tokenConfig[i].provider;
+
+      emit TokenProviderUpdated(_tokenConfig[i].token, address(0), _tokenConfig[i].provider);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    _transferOwnership(_owner);
+    _updateGuardian(_guardian);
+    HYPERNATIVE = _hypernative;
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function setTokenProvider(address _token, address _provider) external onlyOwner {
+    address oldProvider = tokenProvider[_token];
+    tokenProvider[_token] = _provider;
+
+    emit TokenProviderUpdated(_token, oldProvider, _provider);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function deposit(bytes32 _poolId, uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    address[] memory assets;
+    (assets, , ) = VAULT.getPoolTokens(_poolId);
+    uint256 tokenCount = assets.length;
+    (address poolAddress, ) = VAULT.getPool(_poolId);
+
+    if (_tokenAmounts.length != tokenCount) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < tokenCount; ) {
+      uint256 currentBalance = IERC20(assets[i]).balanceOf(address(this));
+
+      if (_tokenAmounts[i] > currentBalance) {
+        revert InsufficientToken(assets[i], currentBalance);
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    for (uint256 i = 0; i < tokenCount; ) {
+      IERC20(assets[i]).safeIncreaseAllowance(address(VAULT), _tokenAmounts[i]);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(
+      WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+      _tokenAmounts,
+      0
```

Comment:
What's this 0?

Reply:
If it's min amount expected, then this is a no-go because we'll get front-run every time and get killed on slippage

Reply:
I think there wouldn't be front-run attack on join and exit pool. anyway. let me find check if there is any method to estimate out amount before. but it would be just waste of fee.

Reply:
@LucasWongC did you find an answer to what the 0 is? there's definitely front-running

Reply:
0 is min amount of bpt token